### ISSUE TITLE
scn.BatchNormalization does not produce results like nn.BatchNorm2d

### ISSUE THREAD

**Issue by jewfro-cuban** [facebookresearch/SparseConvNet#43]
Reproducible code:

```python
import numpy
import sparseconvnet as scn
import torch
from torch import nn

spatial_dim = 2
spatial_shape = (500, 1000)
mask = numpy.zeros(spatial_shape, dtype='uint8')
mask[200:400, 400:600] = 1


dim = 2
inplanes = 3
outplanes = 1
# NOTE: transpose yields indices to be f-contiguous instead of c-contiguous.
# set_locations assumes c-contiguous and converts to c-contiguous when it's not.
# for performance make sure locations (and features) input is c-contiguous
i, j = numpy.where(mask)
n_sites = len(i)
ij = numpy.hstack([i.reshape(-1, 1), j.reshape(-1, 1)])
locations = torch.LongTensor(ij)
features = torch.FloatTensor([[1, 2, 3], ] * n_sites)

sparse_input = scn.InputBatch(spatial_dim, torch.LongTensor(spatial_shape))

sparse_input.add_sample()
assert locations.is_contiguous()
assert features.is_contiguous()
sparse_input.set_locations(locations, features)

dense_input = scn.SparseToDense(dim, inplanes)(sparse_input)


sparse_model = scn.Sequential()
sparse_model.add_module('bn1', scn.BatchNormalization(inplanes, eps=1e-5, momentum=0.1))
sparse_model.add_module('relu1', scn.ReLU())
sparse_model.add_module('conv1', scn.SubmanifoldConvolution(dim, inplanes, outplanes, 1, False))

dense_model = nn.Sequential()
dense_model.add_module('bn1', nn.BatchNorm2d(inplanes, eps=1e-5, momentum=0.1))
dense_model.add_module('relu1', nn.ReLU(inplace=True))
dense_model.add_module('conv1', nn.Conv2d(inplanes, outplanes, kernel_size=1, bias=False))

# set all parameters values to 1 for comparison
for dp, sp in zip(dense_model.parameters(), sparse_model.parameters()):
    sp.data.copy_(torch.ones_like(sp.data))
    dp.data.copy_(torch.ones_like(dp.data))
dense_output = dense_model(dense_input)
sparse_output = sparse_model(sparse_input)


do = dense_output.squeeze().detach()
so = scn.SparseToDense(dim, outplanes)(sparse_output).squeeze().detach()
print("All close:")
print(numpy.allclose(do, so))

# or stack first plane for convinience
stacked_output = numpy.vstack([do, so])
#  matshow(stacked_output)

```

Output is False.
If bn lines are commented out, output becomes True.

Should I expect same outputs with both versions?

**Comment by btgraham** [comment]
Hello. This is actually intended behaviour  (https://arxiv.org/pdf/1706.01307.pdf Section 3, page 4).
SparseConvNet operations totally ignore inactive sites. This allows the spatial_shape to be made arbitrarily large, which can be useful for data augmentation, or irregularly sized input. 
In the dense case, your input contains a mix of [0,0,0]s and [1,2,3]s.
In the sparse case, the [0,0,0]s are discarded, so the mean of the input features is [1,2,3], and the standard deviation is [0,0,0] !

Start the network with a convolution, i.e.
'scn.SubmanifoldConvolution(dim, inplanes, outplanes, 3, False)) - BatchNormReLU - ...'
then the input to subsequent scn.BatchNormalization layers is unlikely to have standard deviation 0.
